Add support for existing Cognito User Pools

pmuens opened the pull request:
This PR adds support for existing Cognito User Pools via a custom resource.

What did you implement:

Closes #4207
Closes #5401
Refs #4241
Refs #5642
Refs #6290

/cc @exoego 

## How did you implement it:

Checks the `existing` config property of the `cognitoUserPool` event and swaps out the Cognito User Pool CloudFormation resource with a custom resource the Serverless Framework provides. The custom resource basically does raw SDK calls to update the existing Cognito User Pool. This way everything is encapsulated via CloudFormation and the stack won't get out of sync if we implement it via raw SDK calls.

## How can we verify it:

```yaml
service: test

provider:
  name: aws
  runtime: nodejs10.x

functions:
  test:
    handler: handler.hello
    events:
      - cognitoUserPool:
          pool: ExistingUserPool
          trigger: CustomMessage
          existing: true
```

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

medikoo commented on lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js:
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+const AWS = require('aws-sdk');
```
We may narrow require to:

```javascript
const Lambda = require('aws-sdk/clients/lambda');
```

While it won't make much difference now. It's a way to reduce size of a lambda, if packaging is setup to take only modules that are required (and I hope in near future we'll have that way integrated into core)

medikoo commented on lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js:
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+const AWS = require('aws-sdk');
```
We may narrow require to:

```javascript
const Lambda = require('aws-sdk/clients/lambda');
```

While it won't make much difference now. It's a way to reduce size of a lambda, if packaging is setup to take only modules that are required (and I hope in near future we'll have that way integrated into core)

pmuens replied:
Great idea :+1:

medikoo commented on lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js:
```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const AWS = require('aws-sdk');
```
Same here

medikoo commented on lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js:
```diff
@@ -0,0 +1,85 @@
+        UserPoolId: userPool.Id,
+      })
+      .promise()
+      .then(data => data)
```
This line doesn't do anything

pmuens replied:
Lol 🤦‍♂

exoego commented on lib/plugins/aws/package/compile/events/cognitoUserPool/index.js:
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
+      if (functionObj.events) {
+        functionObj.events.forEach(event => {
+          if (event.cognitoUserPool) {
+            // return immediately if it's an existing S3 event since we treat them differently
```
typo: S3

exoego commented on lib/plugins/aws/package/compile/events/cognitoUserPool/index.js:
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
+            _.isObject(event.cognitoUserPool) &&
+            event.cognitoUserPool.existing
+          ) {
+            idx = idx += 1;
```
Why not simply `idx++`?

medikoo commented on lib/plugins/aws/package/compile/events/cognitoUserPool/index.js:
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
+        functionObj.events.forEach((event, idx) => {
+          if (
+            event.cognitoUserPool &&
+            _.isObject(event.cognitoUserPool) &&
```
This check is rudimentary

pmuens replied:
🤦‍♂

medikoo commented on tests/integration-all/cognito-user-pool/service/utils.js:
```diff
@@ -0,0 +1,22 @@
+'use strict';
```
This file looks as copy of https://github.com/serverless/serverless/blob/master/tests/integration-all/s3/service/utils.js

Maybe it'll be good to find a way to reuse same file (and just ensure it in all tested services)

pmuens replied:
👍 will do

pmuens commented:
Thanks for the reviews @medikoo and @exoego :+1:

I just addressed your issues and updated the PR accordingly. Other than that I also added support for the usage of multiple custom resources in one service.

pmuens commented:
Thanks for the review @medikoo :+1:

I agree to reduce Lodash usage and I tried to do it here. One problem I ran into was the way `Object.assign` and `_.merge` differ. `Object.assign` creates a copy whereas `_.merge` mutates the object (which is another concern). I needed a way to mutate the object so I used `_.merge` like we do in other places as well.

However I agree that we should definitely try to adhere to vanilla JS as much as possible given that we dropped support for all the old Node.js versions.
